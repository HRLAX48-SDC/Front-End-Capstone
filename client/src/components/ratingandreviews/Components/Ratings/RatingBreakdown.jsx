import React, { useState, useEffect} from 'react';
import { Card, ProgressBar, Stack, Button, OverlayTrigger, Tooltip } from 'react-bootstrap';
import Stars from './Stars.jsx';


function RatingBreakdown (props) {
  const [ratings, setRatings] = useState({});
  const [totalRatings, setTotalRatings] = useState(null);
  const [percentRecommended, setPercentRecommended] = useState(null);
  const [filterList, setFilterList] = useState('')
 
  useEffect(() => {
    setRatings(props.metaData.ratings)
    sumRatings(props.metaData.ratings)
    avgRecommended(props.metaData.recommended)
  },[props.metaData.ratings]);

  const sumRatings = (ratingsObj) => {
    if(props.metaData.ratings){
      let ratingsArr = Object.values(ratingsObj);
      let total = ratingsArr.reduce((a, b) => Number(a) + Number(b));
      setTotalRatings(total);
    }
  }

  const avgRecommended = (recObj) => {
    if(props.metaData.recommended){
      let trueNum = Number(recObj['true']) || 0;
      let falseNum = Number(recObj['false']) || 0;
      let total = trueNum + falseNum;
      let avg = (trueNum / total) * 100;
      let roundAvg = Math.round(avg)
      setPercentRecommended(roundAvg);
    }
  }
  
  const showFilters = (obj) => {
    let filters = '';
    for (let key in obj){
      if(obj[key] === true){
        filters.length < 50 ? filters += ` ${key} stars |` 
        : filters += ` ${key} stars`
      }
    }
    return filters;
  }

  return (
    <Card>
      <Card.Title>
        <Stars ratings={ratings} totalRatings={totalRatings}/>
      </Card.Title>
      <Card.Body>
        { ratings ?
        <>
        <Stack direction="horizontal" gap={2}>
            <div className="percent-recommended">{percentRecommended}%</div> 
            <>of reviewers recommend this product</>
        </Stack>
        <br></br>
        <Stack gap={2}>
          <div className="breakdowns" >
            <OverlayTrigger
              placement="right"
              delay={{ show: 250, hide: 400 }}
              overlay={<Tooltip id="button-tooltip">toggle filter</Tooltip>}
            >
              <Button variant="outline-secondary" size="sm" 
              name="five" onClick={(e) => props.toggleFiltered(e.target.name)}>Five stars: {ratings['5'] || '0'}</Button>
            </OverlayTrigger>
          </div>
            <ProgressBar variant="success" now={(ratings['5'] || 0) / totalRatings * 100} />
          <div className="breakdowns">
            <OverlayTrigger
              placement="right"
              delay={{ show: 250, hide: 400 }}
              overlay={<Tooltip id="button-tooltip">toggle filter</Tooltip>}
            >
              <Button variant="outline-secondary" size="sm"
              name="four" onClick={(e) => props.toggleFiltered(e.target.name)}>Four stars: {ratings['4'] || '0'}</Button>
            </OverlayTrigger>
          </div>
            <ProgressBar variant="success" now={(ratings['4'] || 0) / totalRatings * 100} /> 
          <div className="breakdowns">
            <OverlayTrigger
              placement="right"
              delay={{ show: 250, hide: 400 }}
              overlay={<Tooltip id="button-tooltip">toggle filter</Tooltip>}
            >
              <Button variant="outline-secondary" size="sm"
              name="three" onClick={(e) => props.toggleFiltered(e.target.name)}>Three stars: {ratings['3'] || '0'}</Button>
            </OverlayTrigger>
          </div> 
            <ProgressBar variant="success" now={(ratings['3'] || 0) / totalRatings * 100} />
          <div className="breakdowns">
            <OverlayTrigger
              placement="right"
              delay={{ show: 250, hide: 400 }}
              overlay={<Tooltip id="button-tooltip">toggle filter</Tooltip>}
            >
              <Button variant="outline-secondary" size="sm"
              name="two" onClick={(e) => props.toggleFiltered(e.target.name)}>Two stars: {ratings['2'] || '0'}</Button>
            </OverlayTrigger>
          </div>
            <ProgressBar variant="success" now={(ratings['2'] || 0) / totalRatings * 100} /> 
          <div className="breakdowns">
          <OverlayTrigger
              placement="right"
              delay={{ show: 250, hide: 400 }}
              overlay={<Tooltip id="button-tooltip">toggle filter</Tooltip>}
            >
              <Button variant="outline-secondary" size="sm"
              name="one" onClick={(e) => props.toggleFiltered(e.target.name)}>One star: {ratings['1'] || '0'}</Button>
            </OverlayTrigger>
          </div>
            <ProgressBar variant="success" now={(ratings['1'] || 0) / totalRatings * 100} /> 
          </Stack>
          <br></br>
            {showFilters(props.filteredBy).length > 0 &&
            <>
            <Stack direction="horizontal" gap={2}>
              <h5>Filtered by:</h5>
              <Button variant="outline-secondary" size="sm" 
              className="ms-auto" onClick={() => props.resetFiltered()}>Remove all filters</Button>
            </Stack>
            <div className="filtered-list">{showFilters(props.filteredBy)}</div>
            </>}
          </>
          : <div className="spinner-border" role="status">
              <span className="visually-hidden">Loading...</span>
            </div>
          }
      </Card.Body>
    </Card>
  )
}

export default RatingBreakdown;