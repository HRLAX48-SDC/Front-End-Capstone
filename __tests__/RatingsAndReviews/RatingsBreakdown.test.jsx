import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import RatingBreakdown from '../../client/src/components/ratingandreviews/Components/Ratings/RatingBreakdown.jsx'


describe('<RatingBreakdown />', () =>{
    const metaData = {
        characteristics :{
            Comfort: {id: 142034, value: "3.1627906976744186"},
            Fit: {id: 142032, value: "2.9021739130434783"},
            Length: {id: 142033, value: "2.8045977011494253"},
            Quality: {id: 142035, value: "3.1976744186046512"}
        },
        product_id: "42366",
        ratings: {1: "23", 2: "8", 3: "50", 4: "36", 5: "47"},
        recommended : {false: "68", true: "96"}
    }
    it('should correctly render percent recommended', () => {
        const recommended = renderer
        .create(<RatingBreakdown metaData={metaData.recommended}/>)
        .toJSON();
        expect(recommended).toMatchSnapshot()
    });
    it('should correctly render total reviews', () => {
        const totalRatings = renderer
        .create(<RatingBreakdown metaData={metaData.ratings}/>)
        .toJSON();
        expect(totalRatings).toMatchSnapshot()
    })






});