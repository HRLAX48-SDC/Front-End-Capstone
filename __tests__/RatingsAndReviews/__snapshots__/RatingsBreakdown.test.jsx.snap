// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RatingBreakdown /> should correctly render percent recommended 1`] = `
<div
  className="card"
>
  <div
    className="card-title h5"
  >
    <div
      className="hstack gap-3"
    >
      <div
        className="average-review-header"
      >
        0.0
      </div>
      <span
        style={
          Object {
            "direction": "ltr",
            "display": "inline-block",
            "touchAction": "none",
          }
        }
      >
        <span
          aria-hidden="true"
          className="react-simple-star-rating"
          style={
            Object {
              "cursor": "",
              "display": "inline-block",
              "overflow": "hidden",
              "position": "relative",
              "userSelect": "none",
              "verticalAlign": "middle",
              "whiteSpace": "nowrap",
            }
          }
        >
          <span
            className="empty-icons"
            style={
              Object {
                "color": "#cccccc",
                "display": "inline-block",
              }
            }
          >
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
          </span>
          <span
            className="filled-icons"
            style={
              Object {
                "color": "#ffbc0b",
                "display": "inline-block",
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "top": 0,
                "transition": "",
                "whiteSpace": "nowrap",
                "width": "0%",
              }
            }
            title="0 out of 5"
          >
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
          </span>
        </span>
      </span>
    </div>
    <h5
      className="total-ratings-header"
    >
       ratings
    </h5>
  </div>
  <div
    className="card-body"
  >
    <div
      className="hstack gap-2"
    >
      <div
        className="percent-recommended"
      >
        %
      </div>
      of reviewers recommend this product
    </div>
    <br />
    <div
      className="vstack gap-2"
    >
      <div
        className="breakdowns"
      >
        <button
          className="btn btn-outline-secondary btn-sm"
          disabled={false}
          name="five"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          Five stars: 
          0
        </button>
      </div>
      <div
        className="progress"
      >
        <div
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={NaN}
          className="progress-bar bg-success"
          role="progressbar"
          style={
            Object {
              "width": "NaN%",
            }
          }
        />
      </div>
      <div
        className="breakdowns"
      >
        <button
          className="btn btn-outline-secondary btn-sm"
          disabled={false}
          name="four"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          Four stars: 
          0
        </button>
      </div>
      <div
        className="progress"
      >
        <div
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={NaN}
          className="progress-bar bg-success"
          role="progressbar"
          style={
            Object {
              "width": "NaN%",
            }
          }
        />
      </div>
      <div
        className="breakdowns"
      >
        <button
          className="btn btn-outline-secondary btn-sm"
          disabled={false}
          name="three"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          Three stars: 
          0
        </button>
      </div>
      <div
        className="progress"
      >
        <div
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={NaN}
          className="progress-bar bg-success"
          role="progressbar"
          style={
            Object {
              "width": "NaN%",
            }
          }
        />
      </div>
      <div
        className="breakdowns"
      >
        <button
          className="btn btn-outline-secondary btn-sm"
          disabled={false}
          name="two"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          Two stars: 
          0
        </button>
      </div>
      <div
        className="progress"
      >
        <div
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={NaN}
          className="progress-bar bg-success"
          role="progressbar"
          style={
            Object {
              "width": "NaN%",
            }
          }
        />
      </div>
      <div
        className="breakdowns"
      >
        <button
          className="btn btn-outline-secondary btn-sm"
          disabled={false}
          name="one"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          One star: 
          0
        </button>
      </div>
      <div
        className="progress"
      >
        <div
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={NaN}
          className="progress-bar bg-success"
          role="progressbar"
          style={
            Object {
              "width": "NaN%",
            }
          }
        />
      </div>
    </div>
    <br />
  </div>
</div>
`;

exports[`<RatingBreakdown /> should correctly render total reviews 1`] = `
<div
  className="card"
>
  <div
    className="card-title h5"
  >
    <div
      className="hstack gap-3"
    >
      <div
        className="average-review-header"
      >
        0.0
      </div>
      <span
        style={
          Object {
            "direction": "ltr",
            "display": "inline-block",
            "touchAction": "none",
          }
        }
      >
        <span
          aria-hidden="true"
          className="react-simple-star-rating"
          style={
            Object {
              "cursor": "",
              "display": "inline-block",
              "overflow": "hidden",
              "position": "relative",
              "userSelect": "none",
              "verticalAlign": "middle",
              "whiteSpace": "nowrap",
            }
          }
        >
          <span
            className="empty-icons"
            style={
              Object {
                "color": "#cccccc",
                "display": "inline-block",
              }
            }
          >
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
          </span>
          <span
            className="filled-icons"
            style={
              Object {
                "color": "#ffbc0b",
                "display": "inline-block",
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "top": 0,
                "transition": "",
                "whiteSpace": "nowrap",
                "width": "0%",
              }
            }
            title="0 out of 5"
          >
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={35}
              style={Object {}}
              viewBox="0 0 24 24"
              width={35}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
          </span>
        </span>
      </span>
    </div>
    <h5
      className="total-ratings-header"
    >
       ratings
    </h5>
  </div>
  <div
    className="card-body"
  >
    <div
      className="spinner-border"
      role="status"
    >
      <span
        className="visually-hidden"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
